### ISSUE TITLE
Invalid state error when committing transaction

### ISSUE THREAD

**Issue by roignpar** [fede1024/rust-rdkafka#349]
I ran into this error when trying to commit a producer transaction with version `0.25.0` of rust-rdkafka:
```
Error: KafkaError (Transaction error: EndTxn commit failed: Broker: Producer attempted a transactional operation in an invalid state)
```

Full code to reproduce can be found here: https://github.com/roignpar/rdkafka_tx_commit_error

main.rs:
```rust
use rdkafka::{
    error::KafkaResult,
    producer::{Producer, ThreadedProducer},
    util::Timeout,
    ClientConfig,
};

fn main() -> KafkaResult<()> {
    env_logger::init();

    let mut config = ClientConfig::new();

    config.set("bootstrap.servers", "localhost:19092");
    config.set("transactional.id", "transaction_test");
    config.set("debug", "eos");

    let prod: ThreadedProducer<_> = config.create()?;

    prod.init_transactions(Timeout::Never)?;
    log::info!("Init transactions successful");

    prod.begin_transaction()?;
    log::info!("Begin transaction successful");

    prod.commit_transaction(Timeout::Never)?;
    log::info!("Commit transaction successful");

    Ok(())
}
```

`RUST_LOG=trace cargo run` output:
```
❯ RUST_LOG=trace cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
     Running `target/debug/tx_test_rdkafka`
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: INIT [thrd:app]: librdkafka v1.6.0 (0x10600ff) rdkafka#producer-1 initialized (builtin.features gzip,snappy,sasl,regex,lz4,sasl_plain,plugins, GCC GXX PKGCONFIG INSTALL GNULD LDS C11THREADS LIBDL PLUGINS ZLIB HDRHISTOGRAM SYSLOG SNAPPY SOCKEM CRC32C_HW, debug 0x8000)
[2021-02-18T13:04:07Z TRACE rdkafka::client] Create new librdkafka client 0x56330a6149b0
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: TXNAPI [thrd:app]: Transactional API called: init_transactions
[2021-02-18T13:04:07Z TRACE rdkafka::producer::base_producer] Polling thread loop started
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: TXNSTATE [thrd:main]: Transaction state change Init -> WaitPID
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: IDEMPSTATE [thrd:main]: Idempotent producer state change Init -> RequestPID
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: TXN [thrd:main]: Starting PID FSM timer (fire immediately): Starting idempotent producer
[2021-02-18T13:04:07Z ERROR rdkafka::client] librdkafka: FAIL [thrd:localhost:19092/bootstrap]: localhost:19092/bootstrap: Connect to ipv6#[::1]:19092 failed: Connection refused (after 0ms in state CONNECT)
[2021-02-18T13:04:07Z ERROR rdkafka::client] librdkafka: Global error: BrokerTransportFailure (Local: Broker transport failure): localhost:19092/bootstrap: Connect to ipv6#[::1]:19092 failed: Connection refused (after 0ms in state CONNECT)
[2021-02-18T13:04:07Z ERROR rdkafka::client] librdkafka: Global error: AllBrokersDown (Local: All broker connections are down): 1/1 brokers are down
[2021-02-18T13:04:07Z TRACE rdkafka::producer::base_producer] Received 2 events
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: PIDBROKER [thrd:main]: No brokers available for Transactions (2 broker(s) known)
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: TXNCOORD [thrd:main]: Unable to query for transaction coordinator: No brokers available for Transactions (2 broker(s) known)
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: PIDBROKER [thrd:main]: No brokers available for Transactions (2 broker(s) known)
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: TXNCOORD [thrd:main]: Unable to query for transaction coordinator: No brokers available for Transactions (2 broker(s) known)
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: TXNCOORD [thrd:main]: localhost:19092/1001: FindCoordinator response: Transaction coordinator is broker 1001 (localhost:19092)
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: TXNCOORD [thrd:main]: Transaction coordinator changed from (none) -> localhost:19092/1001: FindCoordinator response
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: IDEMPSTATE [thrd:main]: Idempotent producer state change RequestPID -> WaitTransport
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: TXN [thrd:main]: Starting PID FSM timer: No broker available
[2021-02-18T13:04:07Z ERROR rdkafka::client] librdkafka: FAIL [thrd:TxnCoordinator]: TxnCoordinator: localhost:19092: Connect to ipv6#[::1]:19092 failed: Connection refused (after 0ms in state CONNECT)
[2021-02-18T13:04:07Z ERROR rdkafka::client] librdkafka: Global error: BrokerTransportFailure (Local: Broker transport failure): TxnCoordinator: localhost:19092: Connect to ipv6#[::1]:19092 failed: Connection refused (after 0ms in state CONNECT)
[2021-02-18T13:04:07Z TRACE rdkafka::producer::base_producer] Received 1 events
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: COORD [thrd:main]: TxnCoordinator/1001: Transaction coordinator is now up
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: GETPID [thrd:main]: TxnCoordinator/1001: Acquiring ProducerId
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: IDEMPSTATE [thrd:main]: Idempotent producer state change WaitTransport -> WaitPID
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: GETPID [thrd:main]: Acquired PID{Id:1000,Epoch:14}
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: IDEMPSTATE [thrd:main]: Idempotent producer state change WaitPID -> Assigned
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: TXNSTATE [thrd:main]: Transaction state change WaitPID -> ReadyNotAcked
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: TXNAPI [thrd:app]: Transactional API called: init_transactions
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: TXNSTATE [thrd:main]: Transaction state change ReadyNotAcked -> Ready
[2021-02-18T13:04:07Z INFO  tx_test_rdkafka] Init transactions successful
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: TXNSTATE [thrd:main]: Transaction state change Ready -> InTransaction
[2021-02-18T13:04:07Z INFO  tx_test_rdkafka] Begin transaction successful
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: TXNAPI [thrd:app]: Transactional API called: commit_transaction (begin)
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: TXNSTATE [thrd:main]: Transaction state change InTransaction -> BeginCommit
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: TXNCOMMIT [thrd:app]: Flushing 0 outstanding message(s) prior to commit
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: TXNCOMMIT [thrd:app]: Transaction commit message flush complete
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: TXNAPI [thrd:app]: Transactional API called: commit_transaction
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: TXNSTATE [thrd:main]: Transaction state change BeginCommit -> CommittingTransaction
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: ENDTXN [thrd:main]: EndTxn returned INVALID_TXN_STATE in state CommittingTransaction (may_retry=true)
[2021-02-18T13:04:07Z ERROR rdkafka::client] librdkafka: TXNERR [thrd:main]: Fatal transaction error: Failed to end transaction: Broker: Producer attempted a transactional operation in an invalid state (INVALID_TXN_STATE)
[2021-02-18T13:04:07Z ERROR rdkafka::client] librdkafka: FATAL [thrd:main]: Fatal error: Broker: Producer attempted a transactional operation in an invalid state: Failed to end transaction: Broker: Producer attempted a transactional operation in an invalid state
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: TXNSTATE [thrd:main]: Transaction state change CommittingTransaction -> FatalError
[2021-02-18T13:04:07Z ERROR rdkafka::client] librdkafka: Global error: Fatal (Local: Fatal error): Fatal error: Broker: Producer attempted a transactional operation in an invalid state: Failed to end transaction: Broker: Producer attempted a transactional operation in an invalid state
[2021-02-18T13:04:07Z TRACE rdkafka::producer::base_producer] Destroy ThreadedProducer
[2021-02-18T13:04:07Z TRACE rdkafka::producer::base_producer] Stopping polling
[2021-02-18T13:04:07Z TRACE rdkafka::producer::base_producer] Waiting for polling thread termination
[2021-02-18T13:04:07Z TRACE rdkafka::producer::base_producer] Received 1 events
[2021-02-18T13:04:07Z TRACE rdkafka::producer::base_producer] Polling thread loop terminated
[2021-02-18T13:04:07Z TRACE rdkafka::producer::base_producer] Polling stopped
[2021-02-18T13:04:07Z TRACE rdkafka::producer::base_producer] ThreadedProducer destroyed
[2021-02-18T13:04:07Z TRACE rdkafka::util] Destroying client: 0x56330a6149b0
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: DESTROY [thrd:app]: Terminating instance (destroy flags NoConsumerClose (0x8))
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: IDEMPSTATE [thrd:main]: Idempotent producer state change Assigned -> Terminate
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: DESTROY [thrd:main]: Destroy internal
[2021-02-18T13:04:07Z DEBUG rdkafka::client] librdkafka: DESTROY [thrd:main]: Removing all topics
[2021-02-18T13:04:07Z TRACE rdkafka::util] Destroyed client: 0x56330a6149b0
Error: KafkaError (Transaction error: EndTxn commit failed: Broker: Producer attempted a transactional operation in an invalid state)
[2021-02-18T13:04:07Z TRACE rdkafka::util] Destroying error: 0x7f9dd4002b90
[2021-02-18T13:04:07Z TRACE rdkafka::util] Destroyed error: 0x7f9dd4002b90
```
Using the `bitnami/kafka:2.7.0` image for the broker.

**Comment by benesch** [comment]
Looks like you're trying to commit a transaction with no records in it. That's not permitted by librdkafka (and maybe Kafka itself? Not sure.). In any case, make sure to call `producer.send` at least once after calling `begin_transaction` but before calling `commit_transaction`!

**Comment by roignpar** [comment]
@benesch Thank you for replying.
This is no longer an issue with `rust-rdkafka v0.26.0`.
Yous suggestion did fix things when using `0.25.0`.

**Comment by benesch** [comment]
Oh, great! They must have fixed it upstream.
